warp/Network/Wai/Handler/Warp/MultiMap.hs

```diff
@@ -45,29 +46,22 @@ isEmpty (MultiMap mm) = I.null mm
 
 -- | O(1)
 singleton :: FilePath -> v -> MultiMap v
-singleton path v = MultiMap mm
-  where
-    !h = hash path
-    !mm = I.singleton h [(path,v)]
+singleton path v = MultiMap $ I.singleton (hash path) [(path,v)]
 
 ----------------------------------------------------------------
 
--- | O(N)
+-- | O(M) where M is the number of entries per file
 lookup :: FilePath -> MultiMap v -> Maybe v
-lookup path (MultiMap mm) = case I.lookup h mm of
+lookup path (MultiMap mm) = case I.lookup (hash path) mm of
     Nothing -> Nothing
     Just s  -> Prelude.lookup path s
-  where
-    !h = hash path
 
 ----------------------------------------------------------------
 
 -- | O(log n)
 insert :: FilePath -> v -> MultiMap v -> MultiMap v
-insert path v (MultiMap mm) = MultiMap mm'
-  where
-    !h = hash path
-    !mm' = I.insertWith (<>) h [(path,v)] mm
+insert path v (MultiMap mm) = MultiMap
+  $ I.insertWith (flip (<>)) (hash path) [(path,v)] mm
```

Comment:
I don't understand why `flip` is introduced here.

Reply:
insertWith (++) is quadratic because it appends an element to a list. Does anything depend on the order of inner lists?

Reply:
Isn't `insertWith (++)` *prepending*?
https://hackage.haskell.org/package/containers-0.6.5.1/docs/Data-IntMap-Strict.html#v:insertWith

Reply:
Ah ok, I misunderstood then